File: _Code/TranslationToCpp/GA.cpp

#include "GA.h"

void GA::Tournament(std::vector<Dungeon*> pop, int &parent1, int &parent2) {
    std::unordered_set<int> posHash;
    std::vector<int> parentPosL;
    do {
        int pos = Constants::Next(pop.size());
        // std::cout << "Select: " << pos << std::endl;
        if (posHash.insert(pos).second) {
            parentPosL.push_back(pos);
        }
    } while (posHash.size() != 4);
    // std::cout << "Parent Selection " << pop[parentPosL[0]]->fitness << " <" <<  pop[parentPosL[1]]->fitness << std::endl;
    // std::cout << "Parent Selection " << pop[parentPosL[2]]->fitness << " <" <<  pop[parentPosL[3]]->fitness << std::endl;
    parent1 = pop[parentPosL[0]]->fitness < pop[parentPosL[1]]->fitness ? parentPosL[0] : parentPosL[1];
    parent2 = pop[parentPosL[2]]->fitness < pop[parentPosL[3]]->fitness ? parentPosL[2] : parentPosL[3];
}

void GA::Crossover(Dungeon*& indOriginal1, Dungeon*& indOriginal2) {
    Dungeon *ind1 = nullptr, *ind2 = nullptr;
    Room *roomCut1, *roomCut2;
    std::vector<Room*> failedRooms;
    int prob = Constants::Next(100);
    std::vector<int> specialRooms1, specialRooms2;
    std::vector<int> newSpecial1, newSpecial2;
    int nRooms1 = 0, nRooms2 = 0;
    int nSpecial1 = 0, nSpecial2 = 0;
    int newNSpecial1 = 0, newNSpecial2 = 0;
    bool isImpossible = false;
    bool cleanInd = false;
    std::vector<Room*>::iterator it;

    if (prob < Constants::CROSSOVER_RATE) {
        do {
            if (cleanInd) {
                delete ind1;
                delete ind2;
                ind1 = ind2 = nullptr; // Ensure pointers are null after deletion
            }

            ind1 = indOriginal1->Copy();
            ind2 = indOriginal2->Copy();
            cleanInd = true;

            int cutPosition = Constants::Next(1, ind1->roomList.size());
            roomCut1 = ind1->roomList[cutPosition];
            specialRooms1 = FindNKLR(roomCut1, &nRooms1, &nSpecial1);

            failedRooms.clear();
            do {
                do {
                    cutPosition = Constants::Next(1, ind2->roomList.size());
                    roomCut2 = ind2->roomList[cutPosition];
                    it = std::find_if(failedRooms.begin(), failedRooms.end(),
                                      [roomCut2](Room* room) { return room->Equal(roomCut2); });
                } while (it != failedRooms.end());

                failedRooms.push_back(roomCut2);
                if (failedRooms.size() == ind2->roomList.size() - 1)
                    isImpossible = true;
                specialRooms2 = FindNKLR(roomCut2, &nRooms2, &nSpecial2);
            } while ((nSpecial2 > nRooms1 || nSpecial1 > nRooms2) && !isImpossible);

            if (!isImpossible) {
                ChangeChildren(roomCut1, roomCut2);
                ChangeChildren(roomCut2, roomCut1);

                // Swap the parents of roomCut1 and roomCut2
                Room* tempParent = roomCut1->Parent;
                roomCut1->Parent = roomCut2->Parent;
                roomCut2->Parent = tempParent;

                ind1->RemoveFromGrid(roomCut1);
                ind2->RemoveFromGrid(roomCut2);

                // Swap positions and other properties
                int x = roomCut1->X;
                int y = roomCut1->Y;
                Constants::Direction dir = roomCut1->parentDirection;
                int rotation = roomCut1->rotation;

                roomCut1->X = roomCut2->X;
                roomCut1->Y = roomCut2->Y;
                roomCut1->parentDirection = roomCut2->parentDirection;
                roomCut1->rotation = roomCut2->rotation;

                roomCut2->X = x;
                roomCut2->Y = y;
                roomCut2->parentDirection = dir;
                roomCut2->rotation = rotation;

                ind1->RefreshGrid(roomCut2);
                ind2->RefreshGrid(roomCut1);

                newSpecial1.clear();
                newSpecial2.clear();

                newSpecial2 = FindNKLR(roomCut2, &nRooms2, &newNSpecial2);
                newSpecial1 = FindNKLR(roomCut1, &nRooms1, &newNSpecial1);
            }
        } while ((newNSpecial1 != nSpecial1 || newNSpecial2 != nSpecial2 || nSpecial1 > nRooms2 || nSpecial2 > nRooms1) && !isImpossible);

        if (!isImpossible) {
            roomCut2->FixBranch(specialRooms1);
            roomCut1->FixBranch(specialRooms2);

            ind1->FixRoomList();
            ind2->FixRoomList();

            delete indOriginal1;
            delete indOriginal2;

            indOriginal1 = ind1;
            indOriginal2 = ind2;
            ind1 = ind2 = nullptr; // Avoid double deletion
        }
    }
    delete ind1; // Safe deletion if not null
    delete ind2; // Safe deletion if not null
}


void GA::Mutation(Dungeon* dun) {
    try {
        // Mutate keys, adding or removing a pair
        bool willMutate = Constants::Next(101) <= Constants::MUTATION_RATE;
        MutationOp op;
        if (willMutate) {
			
            op = Constants::Next(101) <= Constants::MUTATION0_RATE ? MutationOp::insertChild : MutationOp::removeLeaf;
            switch (op) {
                case MutationOp::insertChild:
					// std::cout << "ALK" << std::endl;
                    dun->AddLockAndKey();
                    break;
                case MutationOp::removeLeaf:
					// std::cout << "RlK" << std::endl;
                    dun->RemoveLockAndKey();
                    break;
            }
            dun->FixRoomList();
        }

    } catch (const std::exception& e) {
        throw e;
    }
}


void GA::ChangeChildren(Room* cut1, Room* cut2) {
    Room* parent = cut1->Parent;
	
    try {
        switch (cut1->parentDirection) {
            case Constants::Direction::right:
                (*parent).rightChild = cut2;
                break;
            case Constants::Direction::down:
                (*parent).bottomChild = cut2;
                break;
            case Constants::Direction::left:
                (*parent).leftChild = cut2;
                break;
            default:
                std::cout << "Something went wrong in crossover!." << std::endl;
                std::cout << "Direction not supported:\n\tOnly Right, Down, and Left are allowed.\n\n" << std::endl;
                break;
        }
    } catch (std::exception &e) {
        std::cout << "Something went wrong while changing the children!" << std::endl;
        throw e;
    }
}

std::vector<int>  GA::FindNKLR(Room *root, int *nRooms, int *specialCount) {
    std::queue<Room*> toVisit;
    std::vector<int> specialRooms;
    toVisit.push(root);
    *nRooms = 0;
	*specialCount = 0;
	
    while (!toVisit.empty()) {
        (*nRooms)++;
        Room *actualRoom = toVisit.front();
        toVisit.pop();
        RoomType type = actualRoom->type;
        if (type == RoomType::key){
            specialRooms.push_back(actualRoom->keyToOpen);
			(*specialCount)++;
		}
        else if (type == RoomType::locked){
            specialRooms.push_back(-actualRoom->keyToOpen);
			(*specialCount)++;
		}
        if (actualRoom->leftChild != nullptr){
            toVisit.push(actualRoom->leftChild);
		}
        if (actualRoom->bottomChild != nullptr){
            toVisit.push(actualRoom->bottomChild);
		}
        if (actualRoom->rightChild != nullptr){
            toVisit.push(actualRoom->rightChild);
		}
    }
	return specialRooms;
}




float GA::Fitness(Dungeon* ind, int nV, int nK, int nL, float lCoef, int matrixOffset) {
        AStar astar;
		DFS dfs1;
        DFS dfs2;
        DFS dfs3;
        float indSym = 0;
        float fitness = 0;

        if (ind->nLocks > 0) {
			
			float avgUsedRooms = 0;
			
            ind->neededLocks = astar.FindRoute(ind, matrixOffset);
            // std::cout << "Needed locks " << ind->neededLocks << std::endl;

			// Between 3 we calculate the avg
			
			// std::cout << "dfs ... " << Constants::currentValue << std::endl;
			avgUsedRooms += dfs1.FindRoute(ind);
            
			// std::cout << "dfs1 " << avgUsedRooms  << " | " << Constants::currentValue << std::endl;
			
            avgUsedRooms += dfs2.FindRoute(ind);
			// std::cout << "dfs2 " << avgUsedRooms  << " | " << Constants::currentValue << std::endl;
			avgUsedRooms += dfs3.FindRoute(ind);
			//std::cout << "dfs3 " << avgUsedRooms  << " | " << Constants::currentValue << std::endl;
			
            avgUsedRooms = round((avgUsedRooms / 3.0f) * 10000.0)/10000.0;
			
			
            if (ind->neededRooms > ind->roomList.size()) {
                std::cout << "SOMETHING IS REALLY WRONG! Nrooms: " << std::to_string(ind->roomList.size()) << "  Used: " << std::to_string(ind->neededRooms)<< std::endl;
            }
            if (ind->neededLocks > ind->nLocks) {
                std::cout << "SOMETHING IS REALLY WRONG!"<< std::endl;
            }
			
			// std::cout << "needed Locks ... " <<  (2 * (std::abs(nV - ((int)ind->roomList.size())) + std::abs(nK - ind->nKeys) + std::abs(nL - ind->nLocks) + (std::abs(lCoef - ind->avgChildren)*5))) << std::endl;
			// std::cout << " Other " << (ind->nLocks - ind->neededLocks) + std::abs(ind->roomList.size() * 0.8 - avgUsedRooms)<< std::endl;
			
			fitness = (2 * (std::abs(nV - ((int)ind->roomList.size())) + std::abs(nK - ind->nKeys) + std::abs(nL - ind->nLocks) + (std::abs(lCoef - ind->avgChildren)*5))) ;
			fitness += (ind->nLocks - ind->neededLocks) + std::abs(ind->roomList.size() * 0.8 - avgUsedRooms);
			// std::cout << "Final Fitness: " << fitness << std::endl;
			
        } else {
			fitness = (2 * (std::abs(nV - ((int)ind->roomList.size())) + std::abs(nK - ind->nKeys) + std::abs(nL - ind->nLocks) +
					std::abs(lCoef - ind->avgChildren)));
            // std::cout << "Final Fitness NL: " << fitness << std::endl;
        }

		return round(fitness * 10000.0)/10000.0;
    }

std::vector<Dungeon*> GA::CreateDungeon() {
    std::vector<Dungeon*> dungeons;
    

    for (int dungeonID = 0; dungeonID < Constants::POP_SIZE; dungeonID++) {
        Dungeon* individual = new Dungeon();
        individual->GenerateRooms();
        dungeons.push_back(individual->Copy());
    }

    if (testingMode) {
        std::cout << "Start To evolve with first fitness !: " << Constants::currentValue << std::endl;
    }

    for (int gen = 0; gen < Constants::GENERATIONS; gen++) {
		std::vector<Dungeon*> nexGeneration;
		Dungeon* aux;
		
		
        if (testingMode) {
            // std::cout << "GEN " << gen << std::endl;
        }

        int counter = 0;
        for (Dungeon* dungeon : dungeons) {
            if (testingMode) {
                
                // std::cout << "Dungeon: " << counter << std::endl;
                for (Room* r : dungeon->roomList) {
                    // std::cout << r->roomId << " [" << r->X << "," << r->Y << " ]" << r->keyToOpen << std::endl;
                }
            }
            
            float fitness = Fitness(dungeon, Constants::nV, Constants::nK, Constants::nL, Constants::lCoef, Constants::MATRIXOFFSET);
            dungeon->fitness = fitness;

            if(testingMode){
                // std::cout << fitness << std::endl;
            }
            counter++;
        }

		
        float auxFitness = dungeons[0]->fitness;
        int id = 0;
        for (int j = 0; j < dungeons.size(); j++) {
            // std::cout << auxFitness << ">" << dungeons[j]->fitness << ((auxFitness > dungeons[j]->fitness)? "True":"False") << std::endl;
            if (auxFitness > dungeons[j]->fitness) {
                auxFitness = dungeons[j]->fitness;
                id = j;
            }
        }
        aux = dungeons[id]->Copy();
		
		
		// std::cout << "Prepare next gen " << std::endl;
		// NEXT GENERATION Preparation !
		for (int i = 0; i < Constants::POP_SIZE/2; i++)
		{
			// Select parents
			int parentA = 0;
			int parentB = 1;
			
			Tournament(dungeons, parentA, parentB);

			Dungeon* parent1 = dungeons[parentA]->Copy();
			Dungeon* parent2 = dungeons[parentB]->Copy();

			// Evolution !!!
			if ( testingMode){
				// std::cout << "Cross: " << parentA << " | " << parentB << " :::: " << Constants::currentValue << std::endl;
			}
			
			Crossover(parent1, parent2);
            if ( testingMode){
				// std::cout << "End Crossover: " << Constants::currentValue << std::endl;
			}
			Mutation(parent1);
            if ( testingMode){
				// std::cout << "End Mutation 1: " << Constants::currentValue << std::endl;
			}
			Mutation(parent2);
			if ( testingMode){
				// std::cout << "End Mutation 2: " << Constants::currentValue << std::endl;
			}

			parent1->FixRoomList();
			parent2->FixRoomList();
			
			parent1->CalcAvgChildren();
			parent2->CalcAvgChildren();
			
			nexGeneration.push_back(parent1->Copy());
			nexGeneration.push_back(parent2->Copy());

			delete parent1;
			delete parent2;
		}

		// Clean current generation
		while (!dungeons.empty()) {
			Dungeon* dungeon = dungeons.back(); // or front() depending on your needs
			dungeons.pop_back();
			delete dungeon; // Delete the object
		}
	
		
		// Prepare the next generation
		for(Dungeon* d : nexGeneration){
			dungeons.push_back(d);
		}
		
		// Elitism
		dungeons[0]->CleanDungeon();
		delete dungeons[0];
		dungeons[0] = aux;
    }
	
    float auxFitness = dungeons[0]->fitness;
    int id = 0;
    std::cout << "FINAL DUNGEONS" << std::endl << std::endl;
    for (int j = 0; j < dungeons.size(); j++) {
        float fitness = Fitness(dungeons[j], Constants::nV, Constants::nK, Constants::nL, Constants::lCoef, Constants::MATRIXOFFSET);
        dungeons[j]->fitness = fitness;

        if (testingMode) {
            std::cout << fitness << std::endl;

            for (Room* r : dungeons[j]->roomList) {
                std::cout << r->roomId << " [" << r->X << "," << r->Y << " ]" << ((int)r->type) << ": " << r->keyToOpen << std::endl;
            }
        }

        if (auxFitness < dungeons[j]->fitness) {
            auxFitness = dungeons[j]->fitness;
            id = j;
        }
    }
	

    return dungeons;
}


// to test
void GA::ReadSequence(std::string fileName){
	Constants::isForced = true;
	// read file
	std::ifstream ifs(fileName);
    std::vector<int> intVector;
    std::string line;
	Constants::forcedRandomSequence.clear();
	
    if (ifs.is_open()) {
        while (std::getline(ifs, line)) {
            // Convert the line to an integer and save it in the vector
            int number = std::stoi(line);
			//std::cout << number << std::endl;
            Constants::forcedRandomSequence.push_back(number);
        }
        ifs.close();
    } else {
        std::cerr << "Unable to open file" << std::endl;
        Constants::isForced = false;
    }
}

File: Assets/Scripts/LevelGenerator/GA.cs
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;

namespace LevelGenerator
{
    class GA
    {
        /**
         * Locks for multi-thread operations
         */
        private readonly object dfsLock = new object();
        private readonly object avgRoomLock = new object();

        /**
         * Enum to choose between mutation operations
         */
        public enum MutationOp
        {
            insertChild = 0,
            removeLeaf = 1
        };
        /**
         * Search the tree of rooms to find the number of special rooms. 
         * The key room is saved in the list with its positive ID, while the locked room with its negative value of the ID
         */
        List<int> seedDFS = new List<int>();
        static private void FindNKLR(ref int nRooms, ref List<int> specialRooms, Room root)
        {
            Queue<Room> toVisit = new Queue<Room>();
            specialRooms = new List<int>();
            toVisit.Enqueue(root);
            nRooms = 0;
            while (toVisit.Count > 0)
            {
                nRooms++;
                Room actualRoom = toVisit.Dequeue() as Room;
                Type type;
                type = actualRoom.Type;
                if (type == Type.key) {
                    specialRooms.Add(actualRoom.KeyToOpen);
                }
                else if (type == Type.locked) {
                    specialRooms.Add(-actualRoom.KeyToOpen);
                }
                if (actualRoom.LeftChild != null) {
                    toVisit.Enqueue(actualRoom.LeftChild);
                }
                if (actualRoom.BottomChild != null) {
                    toVisit.Enqueue(actualRoom.BottomChild);
                }
                if (actualRoom.RightChild != null) {
                    toVisit.Enqueue(actualRoom.RightChild);
                }
            }
        }

        /**
         * Changes the selected rooms between the parent dungeons
         * To do so, changes in the parent who is their child to the corresponding node
         */
        static public void ChangeChildren(ref Room cut1, ref Room cut2)
        {
            Room parent = cut1.Parent;
            //Check which child is the cut Room (Right, Bottom, Left)
            try
            {
                switch (cut1.ParentDirection)
                {
                    case Util.Direction.right:
                        parent.RightChild = cut2;
                        break;
                    case Util.Direction.down:
                        parent.BottomChild = cut2;
                        break;
                    case Util.Direction.left:
                        parent.LeftChild = cut2;
                        break;
                    default:
                        Debug.Log("Something went wrong in crossover!.\n");
                        Debug.Log("Direction not supported:\n\tOnly Right, Down and Left are allowed.\n\n");
                        break;
                }
            }
            catch(System.Exception e)
            {
                Debug.Log("Something went wrong while changing the children!");
                throw e;
            }
        }
        /*
         * Choose a random room to switch between the parents and arrange every aspect of the room needed after the change
         * Including the grid, and also the exceptions where the new nodes overlap the old ones
         */
        static public void Crossover(ref Dungeon indOriginal1, ref Dungeon indOriginal2)
        {
            Dungeon ind1, ind2;
            //The root of the branch that will be traded
            Room roomCut1, roomCut2;
            //List of rooms that were the root of the branch and led to an impossible crossover (Tabu List)
            List<Room> failedRooms;
            int prob = Util.Next(100);
            //List of special rooms in the branch to be traded of each parent
            List<int> specialRooms1 = new List<int>(), specialRooms2 = new List<int>();
            //List of special rooms in the traded brach after the crossover
            List<int> newSpecial1 = new List<int>(), newSpecial2 = new List<int>();
            //Total number of rooms in each branch that will be traded
            int nRooms1 = 0, nRooms2 = 0;
            //Answers if the trade is possible or not
            bool isImpossible = false;
            if (prob < Constants.CROSSOVER_RATE)
            {
                do
                {
                    ind1 = indOriginal1.Copy();
                    ind2 = indOriginal2.Copy();


                    //Get a random node from the parent, find the number of keys, locks and rooms and add it to the list of future failed rooms
                    int cutPosition = Util.Next(1, ind1.RoomList.Count);
                    // Debug.Log("Cut position " + cutPosition);
                    roomCut1 = ind1.RoomList[cutPosition];
                    FindNKLR(ref nRooms1, ref specialRooms1, roomCut1);
                    // Debug.Log("R1 = (" + nRooms1 + "," + specialRooms1.Count + ")" + " [" + roomCut1.RoomId + "]");
                    failedRooms = new List<Room>();

                    
                    //While the number of Keys and Locks from a branch is greater than the number of rooms of the other branch,
                    //Redraw the cut point (root of the branch).
                    do
                    {
                        do
                        {
                            int randomID = Util.Next(1, ind2.RoomList.Count);
                            roomCut2 = ind2.RoomList[randomID];
                            // Debug.Log("3+ " + roomCut2.RoomId + " | " + failedRooms.Contains(roomCut2) + "    "+ Util.randomSequence.Count);
                        } while (failedRooms.Contains(roomCut2));
                        failedRooms.Add(roomCut2);
                        if (failedRooms.Count == ind2.RoomList.Count - 1)
                            isImpossible = true;
                        FindNKLR(ref nRooms2, ref specialRooms2, roomCut2);
                        // Debug.Log("Special Check " + nRooms2 + " | " + specialRooms2.Count + " Test " + (specialRooms2.Count > nRooms1) + "||" +(specialRooms1.Count > nRooms2));
                    } while ((specialRooms2.Count > nRooms1 || specialRooms1.Count > nRooms2) && !isImpossible);
                    // Debug.Log("R2= " + nRooms2);
                    
                    
                    //Changes the children of the parent's and neighbor's nodes to the node of the other branch if it is not an impossible trade
                    if (!isImpossible)
                    {
                        
                        try
                        {
                            ChangeChildren(ref roomCut1, ref roomCut2);
                            ChangeChildren(ref roomCut2, ref roomCut1);
                        }
                        catch (System.Exception e)
                        {
                            throw e;
                        }


                        //Change the parent of each node
                        Room auxRoom;
                        //Changes the parents of the chosen nodes
                        auxRoom = roomCut1.Parent;
                        roomCut1.Parent = roomCut2.Parent;
                        roomCut2.Parent = auxRoom;
                        
                        //Remove the node and their children from the grid of the old dungeon
                        ind1.RemoveFromGrid(roomCut1);
                        ind2.RemoveFromGrid(roomCut2);

                        //Update the position, parent's direction and rotation of both nodes that are switched
                        int x = roomCut1.X;
                        int y = roomCut1.Y;
                        Util.Direction dir = roomCut1.ParentDirection;
                        int rotation = roomCut1.Rotation;
                        roomCut1.X = roomCut2.X;
                        roomCut1.Y = roomCut2.Y;
                        roomCut1.ParentDirection = roomCut2.ParentDirection;
                        roomCut1.Rotation = roomCut2.Rotation;
                        roomCut2.X = x;
                        roomCut2.Y = y;
                        roomCut2.ParentDirection = dir;
                        roomCut2.Rotation = rotation;

                        //Updates the grid with all the new nodes. If any conflicts arise, handle them as in the child creation.
                        //That is, any overlap will make the node and its children cease to exist 
                        ind1.RefreshGrid(ref roomCut2);
                        ind2.RefreshGrid(ref roomCut1);

                        //Find the number of keys, locks and rooms in the newly switched branches
                        newSpecial1 = new List<int>();
                        newSpecial2 = new List<int>();
                        FindNKLR(ref nRooms2, ref newSpecial2, roomCut2);
                        FindNKLR(ref nRooms1, ref newSpecial1, roomCut1);
                    }
                    //Debug.Log("Repair ? ("  + (newSpecial1.Count != specialRooms1.Count) + " || " + (newSpecial2.Count != specialRooms2.Count) + " || " +                         (specialRooms1.Count > nRooms2) + " || " + (specialRooms2.Count > nRooms1) + ") && " + !isImpossible);
                    //If in the new branches there are special rooms missing or the number of special rooms is greater then the number of total rooms, retry
                } while ((newSpecial1.Count != specialRooms1.Count || newSpecial2.Count != specialRooms2.Count || specialRooms1.Count > nRooms2 || specialRooms2.Count > nRooms1) && !isImpossible);
                //If the crossover can be done, do it. If not, don't.
                if (!isImpossible)
                {

                    /*
                    Debug.Log("FIX: ");
                    foreach (Room r in ind1.RoomList)
                    {
                        Debug.Log(r.RoomId + " [" + r.X + ", " + r.Y + "]" + r.KeyToOpen);
                    }

                    Debug.Log("B");
                    foreach (Room r in ind2.RoomList)
                    {
                        Debug.Log(r.RoomId + " [" + r.X + ", " + r.Y + "]" + r.KeyToOpen);
                    }
                    */

                    //Debug.Log("Fix branches 1: " + Util.randomSequence.Count);
                    roomCut2.FixBranch(specialRooms1);
                    // Debug.Log("Current in sequence " + Util.randomSequence.Count);
                    // Debug.Log("Fix branches 2: " + Util.randomSequence.Count);
                    roomCut1.FixBranch(specialRooms2);
                    // Debug.Log("Fixed " + Util.randomSequence.Count);
                    // Debug.Log("Current in sequence " + Util.randomSequence.Count);
                    //Fix the list of rooms

                    ind1.FixRoomList();
                    ind2.FixRoomList();
                    // Debug.Log("Roomlist ?? " + Util.randomSequence.Count);
                    /*
                    Debug.Log("FIXed: ");
                    foreach (Room r in ind1.RoomList)
                    {
                        Debug.Log(r.RoomId + " [" + r.X + ", " + r.Y + "]" + r.KeyToOpen);
                    }

                    Debug.Log("B");
                    foreach (Room r in ind2.RoomList)
                    {
                        Debug.Log(r.RoomId + " [" + r.X + ", " + r.Y + "]" + r.KeyToOpen);
                    }
                    */

                }
                //Make a copy of the individual and finish the crossover
                indOriginal1 = ind1.Copy();
                indOriginal2 = ind2.Copy();
            }
            
        }
        /*
         * Mutates the individual by adding or removing a key pair
         * TODO: Where did the mutation to add or remove leaf nodes went to? 'O.O
         */
        public static void Mutation(ref Dungeon individual)
        {
            try
            {
                //Mutate keys, adding or removing a pair
                bool willMutate = Util.Next(101) <= Constants.MUTATION_RATE;
                MutationOp op;
                if (willMutate)
                {
                    op = Util.Next(101) <= Constants.MUTATION0_RATE ? MutationOp.insertChild : MutationOp.removeLeaf;
                    switch (op)
                    {
                        case MutationOp.insertChild:
                            individual.AddLockAndKey();
                            break;
                        case MutationOp.removeLeaf:
                            individual.RemoveLockAndKey();
                            break;  
                    }
                    individual.FixRoomList();
                }
            
            }
            catch (System.Exception e)
            {
                throw e;
            }
        }
        /**
         * A simple tournament to select both parents.
         * Selects two random members from the population and picks the better one.
         * Do it again and breed both champions.
         */
        static public void Tournament(List<Dungeon>pop, ref int parent1, ref int parent2)
        {
            HashSet<int> posHash = new HashSet<int>();
            List<int> parentPosL = new List<int>();
            do
            {
                int pos = Util.Next(pop.Count);
                if (posHash.Add(pos))
                {
                    parentPosL.Add(pos);
                }
                // Debug.Log(pos);
            } while (posHash.Count != 4);
            // Debug.Log("Parent Compare: " + (float)pop[parentPosL[0]].fitness + " < " + (float)pop[parentPosL[1]].fitness + " = " + ((float)pop[parentPosL[0]].fitness < (float)pop[parentPosL[1]].fitness));
            // Debug.Log("Parent Compare: " + (float)pop[parentPosL[2]].fitness + " < " + (float)pop[parentPosL[3]].fitness + " = " + ((float)pop[parentPosL[2]].fitness < (float)pop[parentPosL[3]].fitness));
            parent1 = (float)pop[parentPosL[0]].fitness < (float)pop[parentPosL[1]].fitness ? parentPosL[0] : parentPosL[1];
            parent2 = (float)pop[parentPosL[2]].fitness < (float)pop[parentPosL[3]].fitness ? parentPosL[2] : parentPosL[3];
        }

        /* 
         * Calculates the fitness function
         * Fitness is based in the number of rooms, number of keys and locks, the linear coefficient and the number of locks used by the A* and DFS
         */
        public float Fitness(Dungeon ind, int nV, int nK, int nL, float lCoef, int matrixOffset)
        {
            float fitness = 0.0f;
            float avgUsedRoom = 0.0f;
            DFS dfs;
            AStar astar = new AStar();
            //Only use the A* and DFS if there is a lock in the dungeon
            if (ind.nLocks > 0)
            {
                //The A* finds the number of locks needed to finish the dungeon using the heuristic that is close to optimal.
                ind.neededLocks = astar.FindRoute(ind, matrixOffset);
                // Debug.Log("Neeeded locks: " + ind.neededLocks);
                //Execute 3 times the DFS to minimize the randomness
                //Execute them in parallel to make things faster
                //The DFS finds the number of rooms needed to finish the dungeon be exploring blindly.
                
                // Debug.Log("DFS ... "+ Util.randomSequence.Count);
                dfs = new DFS(ind);
                dfs.FindRoute();
                avgUsedRoom += dfs.NVisitedRooms;
                
                // Debug.Log("DFS1 " + avgUsedRoom + " | " + Util.randomSequence.Count);
                
                dfs = new DFS(ind);
                dfs.FindRoute();
                avgUsedRoom += dfs.NVisitedRooms;
                // Debug.Log("DFS2 " + avgUsedRoom + " | " + Util.randomSequence.Count);
                dfs = new DFS(ind);
                dfs.FindRoute();
                avgUsedRoom += dfs.NVisitedRooms;
                // Debug.Log("DFS3 " + avgUsedRoom + " | " + Util.randomSequence.Count);
                
                // Main Change, the Algorithm will do it one at a time !
                /*
                if (Constants.modified == false)
                {
                    Parallel.For(0, 3, (i) =>
                    {

                        lock (dfsLock)
                        {
                            dfs[i] = new DFS(ind);
                        }
                        dfs[i].FindRoute(seedDFS[i]);
                        lock (avgRoomLock)
                        {
                            avgUsedRoom += dfs[i].NVisitedRooms;
                        }
                    });
                }
                */

                // Normal For ???
                ind.neededRooms = Mathf.Round((avgUsedRoom / 3.0f)*10000.0f)/10000.0f;

                //If we need more rooms than the rooms that really exist, something is wrong.
                if (ind.neededRooms > ind.RoomList.Count)
                {
                    Debug.Log("SOMETHING IS REALLY WRONG! Nrooms: " + ind.RoomList.Count + "  Used: " + ind.neededRooms);
                    System.Console.ReadKey();
                }
                //Also, if we find more locks that really exist
                if (ind.neededLocks > ind.nLocks)
                    Debug.Log("SOMETHING IS REALLY WRONG!");

                /*
                Debug.Log("PArts of fitness " + (2 * (System.Math.Abs(nV - ind.RoomList.Count) + System.Math.Abs(nK - ind.nKeys) + System.Math.Abs(nL - ind.nLocks)
                        + System.Math.Abs(lCoef - ind.AvgChildren) * 5))
                     + " O " + ((ind.nLocks - ind.neededLocks) + System.Math.Abs(ind.RoomList.Count * 0.8f - ind.neededRooms)));
                */
                fitness= (2 * (System.Math.Abs(nV - ind.RoomList.Count) + System.Math.Abs(nK - ind.nKeys) + System.Math.Abs(nL - ind.nLocks)
                        + System.Math.Abs(lCoef - ind.AvgChildren) * 5) + (ind.nLocks - ind.neededLocks) + System.Math.Abs(ind.RoomList.Count * 0.8f - ind.neededRooms));

            }

            //If there are no locks, the fitness is based only in the map layout
            else
                fitness = (2*(System.Math.Abs(nV - ind.RoomList.Count) + System.Math.Abs(nK - ind.nKeys) + System.Math.Abs(nL - ind.nLocks) + System.Math.Abs(lCoef - ind.AvgChildren)));

            return Mathf.Round(fitness * 10000.0f) / 10000.0f;
        }
    }
}
